src/blocknative.rs

```diff
@@ -0,0 +1,297 @@
+        let cached_response: Arc<Mutex<CachedResponse<Response>>> = Default::default();
+        let cached_response_clone = cached_response.clone();
+
+        //send one request to initially populate the cached response
+        let request = Request { transport, header };
+        match request.gas_price().await {
+            Ok(response) => match cached_response_clone.lock() {
+                Ok(mut data) => {
+                    *data = CachedResponse {
+                        time: Instant::now(),
+                        data: Some(response),
+                    };
+                }
+                Err(e) => {
+                    tracing::warn!(?e, "failed to obtain lock on cached response");
+                    return Err(anyhow!("failed to obtain lock on cached response"));
+                }
+            },
+            Err(e) => {
+                tracing::warn!(?e, "failed to get initial response from blocknative");
+                return Err(anyhow!("failed to get initial response from blocknative"));
+            }
+        }
+
+        //spawn task for updating the cached response every RATE_LIMIT seconds
+        let handle = task::spawn(async move {
+            loop {
+                match request.gas_price().await {
+                    Ok(response) => match cached_response_clone.lock() {
+                        Ok(mut data) => {
+                            *data = CachedResponse {
+                                time: Instant::now(),
+                                data: Some(response),
+                            };
+                        }
+                        Err(e) => tracing::warn!(?e, "failed to obtain lock on cached response"),
+                    },
+                    Err(e) => tracing::warn!(?e, "failed to get response from blocknative"),
+                }
+                tokio::time::sleep(RATE_LIMIT).await;
+            }
+        });
+
+        Ok(Self {
+            cached_response,
+            handle,
+        })
+    }
+}
+
+#[async_trait::async_trait]
+impl GasPriceEstimating for BlockNative {
+    async fn estimate_with_limits(&self, _gas_limit: f64, time_limit: Duration) -> Result<f64> {
+        let cached_response = match self.cached_response.lock() {
+            Ok(cached_response) => cached_response.clone(),
+            Err(e) => {
+                tracing::warn!(?e, "failed to obtain lock on cached response");
+                return Err(anyhow!("failed to obtain lock on cached response"));
+            }
+        };
+
+        estimate_with_limits(time_limit, cached_response)
+    }
+}
+
+fn estimate_with_limits(
+    time_limit: Duration,
+    cached_response: CachedResponse<Response>,
+) -> Result<f64> {
+    if Instant::now().saturating_duration_since(cached_response.time) > CACHED_RESPONSE_VALIDITY {
+        return Err(anyhow!("cached response is stale"));
+    }
+
+    if let Some(block) = cached_response
+        .data
+        .unwrap_or_default()
```

Comment:
Nit: In the CachedResponse struct we're reusing the response is Optional. But in this module we never set it to None (and I don't think we should) which is why you need the unwrap_or_default. If we use our own struct we can get rid of the Option.

Reply:
CachedReponse struct defined for each module. Fixed.